dynatrace_topology/stackstate_checks/dynatrace_topology/dynatrace_topology.py

```diff
@@ -0,0 +1,276 @@
+# (C) StackState 2020
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from stackstate_checks.base import AgentCheck, ConfigurationError, TopologyInstance
+from .dynatrace_exception import DynatraceError
+
+import requests
+from requests import Session
+import yaml
+from datetime import datetime
+
+
+class DynatraceTopologyCheck(AgentCheck):
+
+    INSTANCE_TYPE = "dynatrace"
+    SERVICE_CHECK_NAME = "dynatrace"
+
+    def __init__(self, name, init_config, instances=None):
+        AgentCheck.__init__(self, name, init_config, instances)
+        self.url = None
+        self.token = None
+        self.tags = []
+        self.environment = None
+        self.domain = None
+        self.verify = None
+        self.cert = None
+        self.keyfile = None
+
+    def get_instance_key(self, instance):
+        if 'url' not in instance:
+            raise ConfigurationError('Missing url in configuration.')
+
+        return TopologyInstance(self.INSTANCE_TYPE, instance["url"], with_snapshots=False)
+
+    def check(self, instance):
+        """
+        Integration logic
+        """
+        if 'url' not in instance:
+            raise ConfigurationError('Missing URL in configuration.')
+        if 'token' not in instance:
+            raise ConfigurationError('Missing API Token in configuration.')
+
+        self.url = instance.get('url')
+        self.token = instance.get('token')
+        self.tags = instance.get('tags', [])
+        self.domain = instance.get('domain', '')
+        self.environment = instance.get('environment', 'production')
+        self.verify = instance.get('verify', True)
+        self.cert = instance.get('cert', '')
+        self.keyfile = instance.get('keyfile', '')
+
+        try:
+            self.start_snapshot()
+            self.process_topology()
+            self.stop_snapshot()
+            msg = "Dynatrace topology processed successfully"
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.OK, tags=self.tags, message=msg)
+        except Exception as e:
+            self.log.exception(str(e))
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=self.tags, message=str(e))
+
+    def process_topology(self):
+        """
+        Collects components and relations for each component type from dynatrace smartscape topology API
+        """
+        start_time = datetime.now()
+        self.log.info("Starting the collection of topology")
+        services = self.collect_services()
+        processes = self.collect_processes()
+        processes_groups = self.collect_process_groups()
+        hosts = self.collect_hosts()
+        applications = self.collect_applications()
+        topology = {"service": services, "process": processes, "process-group": processes_groups, "host": hosts,
+                    "application": applications}
+        # process component for each type
+        for comp_type, response in topology.items():
```

Comment:
Since you're only checking whether the response contains an error when processing, you might do a whole bunch of requests to DT before realizing that you've gotten an error. Perhaps you could check for the error in the `get_json_repsonse`? Perhaps call it `get_dynatrace_json_response` then.